src/ray/core_worker/reference_count.cc

```diff
@@ -27,59 +27,55 @@ void ReferenceCounter::AddOwnedObject(const ObjectID &object_id, const TaskID &o
 
 void ReferenceCounter::AddLocalReference(const ObjectID &object_id) {
   absl::MutexLock lock(&mutex_);
-  auto entry = object_id_refs_.find(object_id);
-  if (entry == object_id_refs_.end()) {
+  auto it = object_id_refs_.find(object_id);
+  if (it == object_id_refs_.end()) {
     // TODO: Once ref counting is implemented, we should always know how the
     // ObjectID was created, so there should always be an entry.
-    entry = object_id_refs_.emplace(object_id, Reference()).first;
+    it = object_id_refs_.emplace(object_id, Reference()).first;
   }
-  entry->second.local_ref_count++;
+  it->second.local_ref_count++;
 }
 
 void ReferenceCounter::RemoveLocalReference(const ObjectID &object_id,
                                             std::vector<ObjectID> *deleted) {
   absl::MutexLock lock(&mutex_);
-  auto entry = object_id_refs_.find(object_id);
-  if (entry == object_id_refs_.end()) {
+  auto it = object_id_refs_.find(object_id);
+  if (it == object_id_refs_.end()) {
     RAY_LOG(WARNING) << "Tried to decrease ref count for nonexistent object ID: "
                      << object_id;
     return;
   }
-  if (--entry->second.local_ref_count == 0 &&
-      entry->second.submitted_task_ref_count == 0) {
-    object_id_refs_.erase(entry);
-    deleted->push_back(object_id);
+  if (--it->second.local_ref_count == 0 && it->second.submitted_task_ref_count == 0) {
+    DeleteReferenceInternal(it, deleted);
   }
 }
 
 void ReferenceCounter::AddSubmittedTaskReferences(
     const std::vector<ObjectID> &object_ids) {
   absl::MutexLock lock(&mutex_);
   for (const ObjectID &object_id : object_ids) {
-    auto entry = object_id_refs_.find(object_id);
-    if (entry == object_id_refs_.end()) {
+    auto it = object_id_refs_.find(object_id);
+    if (it == object_id_refs_.end()) {
       // TODO: Once ref counting is implemented, we should always know how the
```

Comment:
Have you tried changing this if to an assertion? It seems like it should work now.

Reply:
This assertion fails in the case of randomly generated object ids, which are used in some tests.

Reply:
Ah actually it isn't failing from randomly generated object ids, but because we don't actually hold a python reference for large arguments that are passed implicitly by reference